### PR TITLE
feat: [CDS-40811]: service now approval change window, v2 endpoint

### PR DESCRIPTION
**Pull request by udayvunnam**
### Summary

- Change window with optional layout
- Required if the checkbox is enabled
- v2 endpoint - `useGetServiceNowIssueCreateMetadataV2` to avoid calling the same endpoint with various query params.

`data: [] -> data: {fields: [], fieldInternalTypeMap: {}}`

#### Screenshots
Kapture 2022-07-29 at 09 29 03

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [x] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Review by sonaligoyal-harness — DISMISSED** [review]

File: src/modules/70-pipeline/components/PipelineSteps/Steps/Common/ApprovalRejectionCriteria.module.scss [hunk 1]
```diff
@@ -117,3 +117,16 @@
   color: var(--grey-800);
   font-weight: 600;
 }
+
+.approvalChangeWindow {
+  margin-top: var(--spacing-xlarge);
+
+  .windowSelection {
```
Comment (yc147) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/Common/ApprovalRejectionCriteria.module.scss:124]: Why not use Layout.Vertical?

File: src/modules/70-pipeline/components/PipelineSteps/Steps/ServiceNowApproval/ServiceNowApprovalChangeWindow.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    disabled: !formik.values?.spec.enableChangeWindow || readonly || fetchingServiceNowTicketTypes,
+    selectItems: fetchingServiceNowTicketTypes
+      ? [{ label: getString('pipeline.serviceNowApprovalStep.fetchingTicketTypesPlaceholder'), value: '' }]
+      : serviceNowTicketTypesOptions,
+
+    placeholder: fetchingServiceNowTicketTypes
+      ? getString('pipeline.serviceNowApprovalStep.fetchingTicketTypesPlaceholder')
+      : serviceNowTicketTypesFetchErrorMessage
+      ? serviceNowTicketTypesFetchErrorMessage
+      : `${getString('select')} ${placeholder}`,
```
Comment (rameshx) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/ServiceNowApproval/ServiceNowApprovalChangeWindow.tsx:45]:
```suggestion
      : serviceNowTicketTypesFetchErrorMessage || `${getString('select')} ${placeholder}`,
```

**Review by PratyushGar — DISMISSED** [review]

File: src/modules/70-pipeline/components/PipelineSteps/Steps/ServiceNowApproval/helper.ts [hunk 3]
```diff
@@ -26,19 +26,22 @@ export const processInitialValues = (values: ServiceNowApprovalData): ServiceNow
...
   return {
     ...values,
     spec: {
-      ...values.spec,
+      ...restOfSpec,
       connectorRef: values.spec.connectorRef,
       approvalCriteria: getApprovalRejectionCriteriaForSubmit(values.spec.approvalCriteria),
-      rejectionCriteria: getApprovalRejectionCriteriaForSubmit(values.spec.rejectionCriteria)
+      rejectionCriteria: getApprovalRejectionCriteriaForSubmit(values.spec.rejectionCriteria),
+      ...(enableChangeWindow ? { changeWindow } : undefined)
```
Comment (yc147) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/ServiceNowApproval/helper.ts:44]: Can it be changed to `...(enableChangeWindow && { changeWindow })` ?

File: src/modules/70-pipeline/components/PipelineSteps/Steps/ServiceNowApproval/types.ts [hunk 4]
```diff
@@ -29,6 +29,11 @@ export interface ServiceNowApprovalData extends StepElementConfig {
     ticketNumber: string
     approvalCriteria: ApprovalRejectionCriteria
     rejectionCriteria: ApprovalRejectionCriteria
+    enableChangeWindow?: boolean
+    changeWindow?: {
+      startField: string
+      endField: string
```
Comment (rameshx) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/ServiceNowApproval/types.ts:35]:
```suggestion
      start: string
      end: string
```

**Review by rameshx — APPROVED** [review]

File: src/modules/70-pipeline/components/PipelineSteps/Steps/ServiceNowApproval/types.ts (hunk 4 above)
Comment (udayvunnam) [reply src/modules/70-pipeline/components/PipelineSteps/Steps/ServiceNowApproval/types.ts:35]: came from backend, I guess the field suffix is unnecessary too. Let me take that up with backend

File: src/modules/70-pipeline/components/PipelineSteps/Steps/Common/ApprovalRejectionCriteria.module.scss (hunk 1 above)
Comment (udayvunnam) [reply src/modules/70-pipeline/components/PipelineSteps/Steps/Common/ApprovalRejectionCriteria.module.scss:124]: had to make margin-top and `padding as  !important when Layout is used, but we can make that work too

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 6 Issues 
* Code issue 6 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/e864badd0b95ab81c060b15b85f15701/pull-request/10285)

**Review by chetanvekariya-harness — APPROVED** [review]

**Review by yc147 — APPROVED** [review]

**Review by sonaligoyal-harness — APPROVED** [review]

**Comment by udayvunnam** [comment]
dockerize

**Review by PratyushGar — APPROVED** [review]

**Review by ishantmehtaharness — APPROVED** [review]
Looks Good

**Pull request closed without merging**